### PR TITLE
[JENKINS-54601] Move security fixes to async-http-client library

### PR DESCRIPTION
**Pull request by varyvol**
[JENKINS-54601](https://issues.jenkins-ci.org/browse/JENKINS-54601)

Move the functionality to the library.

Downstream of https://github.com/jenkinsci/async-http-client/pull/3 

@stephenc @dwnusbaum @batmat @oleg-nenashev

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -85,7 +85,7 @@
     <dependency>
       <groupId>com.ning</groupId>
       <artifactId>async-http-client</artifactId>
-      <version>1.7.24</version>
+      <version>1.7.24-jenkins-1-20181114.155326-3</version>
```
Comment (varyvol) [inline pom.xml]: To be replaced with the release once [async-http-client PR#3](https://github.com/jenkinsci/async-http-client/pull/3) is merged and released.

**Review by fcojfernandez — APPROVED** [review]
LGTM

**Review by jvz — APPROVED** [review]
LGTM.

**Review by MRamonLeon — APPROVED** [review]

File: src/main/java/jenkins/plugins/asynchttpclient/AHCUtils.java [hunk 2]
```diff
@@ -70,93 +58,4 @@ public static ProxyServer getProxyServer() {
...
-        } : new DefaultHostnameVerifier();
-    }
-
-    /**
-     * Return the default {@link SSLContext} to use with {@link AsyncHttpClient}.
-     *
-     * @return the default {@link SSLContext} to use with {@link AsyncHttpClient}.
-     * @since 1.7.24.1
-     */
-    public static SSLContext getSSLContext() {
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/plugins/asynchttpclient/AHCUtils.java]: I have found usages in some proprietary plugins I have access to. I believe this method need a compatible API if possible

File: src/main/java/jenkins/plugins/asynchttpclient/util/DefaultHostnameVerifier.java [hunk 3]
```diff
@@ -1,148 +0,0 @@
...
- * description of hostname verification and why it's important,
- * please read
- * <a href="http://tersesystems.com/2014/03/23/fixing-hostname-verification/">Fixing
- * Hostname Verification</a>.
- * <p>
- * This code is based on Kevin Locke's <a href="http://kevinlocke.name/bits/2012/10/03/ssl-certificate-verification-in-dispatch-and-asynchttpclient/">guide</a> .
- * </p>
- */
-@Restricted(NoExternalUse.class)
-public class DefaultHostnameVerifier implements HostnameVerifier {
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/plugins/asynchttpclient/util/DefaultHostnameVerifier.java:38]: Found some internal usages in a private-source component

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Although I see no usages of the removed utility classes inside `jenkinsci` organization, I see some in private-source components I have access to. I assume that other plugin developers may have been using these APIs as well. I cannot find usages in GitHub tho: https://github.com/search?q=%22plugins.asynchttpclient.util%22&type=Code

I request changes for now so that the potential compatibility impact is reviewed again. Once the impact is acknowledged and documented in the changelog, we can proceed IMHO.

It may also make sense to mark this plugin as incompatible, TBD

**Comment by varyvol** [comment]
@oleg-nenashev I've restored the compatibility just in case someone was using it, since the functionality is identical.

**Review by fcojfernandez — APPROVED** [review]

File: pom.xml (hunk 1 above)
Comment (oleg-nenashev) [inline pom.xml]: I recommend adding link to the location of the custom for in comments

File: src/main/java/jenkins/plugins/asynchttpclient/AHC.java [hunk 4]
```diff
@@ -34,6 +34,7 @@
      */
     @SuppressFBWarnings(value = "MS_SHOULD_BE_FINAL", justification = "Allow runtime modification")
     @Restricted(NoExternalUse.class) // no direct linking against this field please
+    @Deprecated // 1.7.24-jenkins-1 includes this logic already
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/plugins/asynchttpclient/AHC.java]: Add it to javadoc:

```
@deprecated Async HTTP Client  1.7.24-jenkins-1 includes this logic already
```

File: src/main/java/jenkins/plugins/asynchttpclient/AHCUtils.java [hunk 5]
```diff
@@ -76,6 +77,7 @@ public static ProxyServer getProxyServer() {
      * @return the default {@link HostnameVerifier} to use with {@link AsyncHttpClient}.
      * @since 1.7.24.1
      */
+    @Deprecated // 1.7.24-jenkins-1 includes this logic already
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/plugins/asynchttpclient/AHCUtils.java]: same

File: src/main/java/jenkins/plugins/asynchttpclient/AHCUtils.java [hunk 6]
```diff
@@ -91,6 +93,7 @@ public boolean verify(String s, SSLSession sslSession) {
      * @return the default {@link SSLContext} to use with {@link AsyncHttpClient}.
      * @since 1.7.24.1
      */
+    @Deprecated // 1.7.24-jenkins-1 includes this logic already
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/plugins/asynchttpclient/AHCUtils.java]: same

**Review by oleg-nenashev — COMMENTED** [review]
Javadoc + Open comment:

> It may also make sense to mark this plugin as incompatible, TBD

File: src/main/java/jenkins/plugins/asynchttpclient/AHC.java (hunk 4 above)
Comment (varyvol) [reply src/main/java/jenkins/plugins/asynchttpclient/AHC.java]: Changed.

File: src/main/java/jenkins/plugins/asynchttpclient/AHCUtils.java (hunk 5 above)
Comment (varyvol) [reply src/main/java/jenkins/plugins/asynchttpclient/AHCUtils.java]: Changed.

File: src/main/java/jenkins/plugins/asynchttpclient/AHCUtils.java (hunk 6 above)
Comment (varyvol) [reply src/main/java/jenkins/plugins/asynchttpclient/AHCUtils.java]: Changed.

File: pom.xml (hunk 1 above)
Comment (varyvol) [reply pom.xml]: Added

**Comment by varyvol** [comment]
@oleg-nenashev about your open comment: why would it be incompatible?

File: src/main/java/jenkins/plugins/asynchttpclient/AHCUtils.java [hunk 7]
```diff
@@ -75,7 +76,9 @@ public static ProxyServer getProxyServer() {
      *
      * @return the default {@link HostnameVerifier} to use with {@link AsyncHttpClient}.
      * @since 1.7.24.1
+     * @deprecated Async HTTP Client 1.7.24-jenkins-1 includes this logic already
      */
+    @Deprecated // 1.7.24-jenkins-1 includes this logic already
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/plugins/asynchttpclient/AHCUtils.java]: actually no need to keep comments anymore

**Review by oleg-nenashev — APPROVED** [review]
@varyvol It is formally incompatible, because the upstream library includes binary-incompatible changes. Whether it is enough to mark the plugin as incompatible - it is a maintainer's decision. Either way is acceptable to me.

My concerns about APIs are addressed here, hence approved

**Pull request merged**